content/courses/intro-to-embedded-sys-prog/chapters/interacting_with_devices.rst

```diff
@@ -3,152 +3,1633 @@ Interacting with Devices
+:ada:`Atomic` is pertinent too. More in a moment.) As a result, the
+compiler will generate load or store instructions for every read or
+update to the object that occurs in the source code. Furthermore, it
+cannot generate any additional loads or stores to that variable, and it
+cannot reorder loads or stores from their order in the source code.
+"What You See Is What You Get" in other words.
+
+.. code-block:: ada
+
+   with Ada.Text_IO;  use Ada.Text_IO;
+
+   procedure Demo is
+      Output : File_Type;
+      Silly  : Integer with Volatile;
+   begin
+      Silly := 0;
+      Create (Output, Out_File, "output.txt");
+      Put (Output, "42");
+      Close (Output);
+   end Demo;
+
+If we compile the above, even with warnings enabled we won't get any
+because the compiler is now required to generate the assignment for
+:ada:`Silly`.
+
+The variable :ada:`Silly` is not even a memory-mapped object, but
+remember that we said these aspects are important to the tasking context
+too, for shared variables. We're ignoring that context in this course.
+
+Perhaps the issue is that you want to write some Ada code that interacts
+with a memory-mapped device, and you want to have exactly the load and
+store instructions generated that match those of the Ada code. The
+memory-mapped device will be volatile, in other words, but some other
+variables might need to be volatile too.
+
+For example, :doc:`elsewhere <./multi_language_development>`
+we said that the best way to achieve exact assembly instruction
+sequences is the use of machine-code inserts (MCI). That's true, but for
+the moment let's say we want to write it in Ada without the MCIs. Our
+earlier example was the memory-mapped GPIO ports on Arm microcontrollers
+produced by ST Microelectronics. Specifically, these ports have a "lock"
+per GPIO pin that allows the developer to configure the pin and then
+lock it so that no other configuration can accidentally change the
+configuration of that pin. Doing so requires an exact sequence of loads
+and stores. If we wrote this in Ada it would look like this:
+
+.. code-block:: ada
+
+   procedure Lock (Port : in out GPIO_Port;  Pin : GPIO_Pin) is
+      Temp : Word with Volatile;
+   begin
+      --  set the lock control bit and the pin bit, clear the others
+      Temp := LCCK or Pin'Enum_Rep;
+      --  write the lock and pin bits
+      Port.LCKR := Temp;
+      --  clear the lock bit in the upper half
+      Port.LCKR := Pin'Enum_Rep;
+      --  write the lock bit again
+      Port.LCKR := Temp;
+      --  read the lock bit
+      Temp := Port.LCKR;
+      --  read the lock bit again
+      Temp := Port.LCKR;
+   end Lock;
+
+:ada:`Temp` is marked volatile for the sake of getting exactly the load
+and stores that we express in the source code, corresponding to the
+hardware locking protocol.
+
+This high-level coding approach will work, and is simple enough that not
+using MCIs is not unreasonable. However, what really argues against it
```

Comment:
Minor comment... There are three negatives here:

> not using MCIs is not unreasonable.

Maybe we could rewrite this to:

> not using MCIs can be reasonable.

or something similar.

Reply:
How about:

This high-level coding approach will work, and is simple enough that MCIs need not be used.

Reply:
Perfect. :-)